terminal-emulator/src/main/java/com/termux/terminal/TerminalEmulator.java

```diff
@@ -0,0 +1,2447 @@
+                                    value = 0; // 0=not recognized, 3=permanently set, 4=permanently reset
+                                }
+                            }
+                            mSession.write(String.format(Locale.US, "\033[?%d;%d$y", mode, value));
+                        } else {
+                            unknownSequence(b);
+                        }
+                        break;
+                    case ESC_CSI_ARGS_SPACE:
+                        int arg = getArg0(0);
+                        switch (b) {
+                            case 'q': // "${CSI}${STYLE} q" - set cursor style (http://www.vt100.net/docs/vt510-rm/DECSCUSR).
+                                switch (arg) {
+                                    case 0: // Blinking block.
+                                    case 1: // Blinking block.
+                                    case 2: // Steady block.
+                                        mCursorStyle = TERMINAL_CURSOR_STYLE_BLOCK;
+                                        break;
+                                    case 3: // Blinking underline.
+                                    case 4: // Steady underline.
+                                        mCursorStyle = TERMINAL_CURSOR_STYLE_UNDERLINE;
+                                        break;
+                                    case 5: // Blinking bar (xterm addition).
+                                    case 6: // Steady bar (xterm addition).
+                                        mCursorStyle = TERMINAL_CURSOR_STYLE_BAR;
+                                        break;
+                                }
+                                break;
+                            case 't':
+                            case 'u':
+                                // Set margin-bell volume - ignore.
+                                break;
+                            default:
+                                unknownSequence(b);
+                        }
+                        break;
+                    case ESC_CSI_ARGS_ASTERIX:
+                        int attributeChangeExtent = getArg0(0);
+                        if (b == 'x' && (attributeChangeExtent >= 0 && attributeChangeExtent <= 2)) {
+                            // Select attribute change extent (DECSACE - http://www.vt100.net/docs/vt510-rm/DECSACE).
+                            setDecsetinternalBit(DECSET_BIT_RECTANGULAR_CHANGEATTRIBUTE, attributeChangeExtent == 2);
+                        } else {
+                            unknownSequence(b);
+                        }
+                        break;
+                    default:
+                        unknownSequence(b);
+                        break;
+                }
+                if (!mContinueSequence) mEscapeState = ESC_NONE;
+                break;
+        }
+    }
+
+    /** When in {@link #ESC_P} ("device control") sequence. */
+    private void doDeviceControl(int b) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.